File: code-library/03-ols-regression/ols-regression.py
# ==============================================================================
#  MODULE 3: OLS LINEAR REGRESSION
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Stack: statsmodels (OLS, robust SEs), pandas, numpy, matplotlib/seaborn
# ==============================================================================

import pandas as pd
import numpy as np
import statsmodels.formula.api as smf   # R-style formula interface
import statsmodels.api as sm            # direct matrix interface
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

ces = pd.read_parquet("CES2020_clean.parquet")

# Convert categoricals for formula API
# C(var) in statsmodels formula creates dummy variables (like Stata's i.)
# The lowest category is the reference group by default.


# ==============================================================================
# SECTION 1: SIMPLE (BIVARIATE) OLS
# ==============================================================================

# smf.ols(formula, data).fit() mirrors R's lm() formula syntax.
# ~ separates outcome from predictors; + adds predictors.
# missing="drop" does listwise deletion (default behavior).

m_simple = smf.ols("imm_restrict ~ education", data=ces).fit()

print(m_simple.summary())

# Key output:
#   coef     = unstandardized coefficient
#   std err  = standard error
#   t        = t-statistic
#   P>|t|    = two-tailed p-value
#   [0.025, 0.975] = 95% confidence interval
#   R-squared = proportion of Y variance explained


# ==============================================================================
# SECTION 2: MULTIPLE REGRESSION
# ==============================================================================

# C(var) creates dummy variables from a categorical column.
# The reference category is the lowest value (1 = Male, 1 = Northeast).

m_demog = smf.ols(
    "imm_restrict ~ education + age + C(sex) + C(census_region)",
    data=ces
).fit()

print(m_demog.summary())


# ==============================================================================
# SECTION 3: FULL MODEL
# ==============================================================================

m_full = smf.ols(
    "imm_restrict ~ education + age + C(sex) + C(census_region) "
    "+ ideology5 + party_id7 + college",
    data=ces
).fit()

print(m_full.summary())

# Quick coefficient table
coef_df = pd.DataFrame({
    "coef":    m_full.params,
    "se":      m_full.bse,
    "t":       m_full.tvalues,
    "p":       m_full.pvalues,
    "ci_low":  m_full.conf_int()[0],
    "ci_high": m_full.conf_int()[1],
}).round(4)

print(coef_df)


# ==============================================================================
# SECTION 4: COMPARING NESTED MODELS (F-TEST)
# ==============================================================================

# Use statsmodels' f_test() to test whether added coefficients are jointly zero.
# This is the equivalent of Stata's "test ideology5 party_id7" after regress.

# Test H0: ideology5 == 0 AND party_id7 == 0 simultaneously
hypotheses = "ideology5 = 0, party_id7 = 0"
f_result = m_full.f_test(hypotheses)
f_val = float(np.squeeze(f_result.fvalue))
print(f"\nF-test (ideology5 = party_id7 = 0):")
print(f"  F = {f_val:.3f}, p = {f_result.pvalue:.4f}")


# ==============================================================================
# SECTION 5: INTERACTION TERMS
# ==============================================================================

# In statsmodels formula: * creates main effects + interaction
# a*b = a + b + a:b

m_interact = smf.ols(
    "imm_restrict ~ education * C(party_id3) + age + C(sex)",
    data=ces
).fit()

print(m_interact.summary())

# Generate predicted values across education × party combinations
from itertools import product

pred_grid = pd.DataFrame(
    list(product(range(1, 7), [1, 2, 3])),
    columns=["education", "party_id3"]
)
pred_grid["age"] = ces["age"].mean()
pred_grid["sex"] = 1   # reference category

pred_grid["predicted"] = m_interact.predict(pred_grid)

# Plot
fig, ax = plt.subplots(figsize=(9, 5))
colors = {1: "blue", 2: "red", 3: "darkgreen"}
labels = {1: "Democrat", 2: "Republican", 3: "Independent"}

for pid, grp in pred_grid.groupby("party_id3"):
    ax.plot(grp["education"], grp["predicted"],
            color=colors[pid], label=labels[pid], linewidth=2, marker="o")

ax.set_xticks(range(1, 7))
ax.set_xticklabels(["No HS","HS","Some coll.","Assoc.","Bach.","Postgrad"])
ax.set_xlabel("Education Level")
ax.set_ylabel("Predicted Restrictionism (0–5)")
ax.set_title("Predicted Immigration Restrictionism by Education & Party")
ax.legend()
plt.tight_layout()
plt.savefig("interact_imm_educ_party.png", dpi=300)
plt.close()


# ==============================================================================
# SECTION 6: ROBUST STANDARD ERRORS
# ==============================================================================

# fit().get_robustcov_results(cov_type="HC3") applies Huber-White robust SEs.
# cov_type options:
#   "HC0" "HC1" "HC2" "HC3" — heteroskedasticity-robust (HC3 is most conservative)
#   "cluster" — clustered SEs (specify groups= argument)

m_robust = smf.ols(
    "imm_restrict ~ education + age + ideology5 + C(sex) + C(census_region)",
    data=ces
).fit(cov_type="HC3")

print(m_robust.summary())

# Compare: standard vs. robust SEs for education coefficient
m_standard = smf.ols(
    "imm_restrict ~ education + age + ideology5 + C(sex)",
    data=ces
).fit()

m_hc3 = smf.ols(
    "imm_restrict ~ education + age + ideology5 + C(sex)",
    data=ces
).fit(cov_type="HC3")

print(f"\nEducation coef: {m_standard.params['education']:.4f}")
print(f"  Standard SE:  {m_standard.bse['education']:.4f}")
print(f"  Robust SE:    {m_hc3.bse['education']:.4f}")


# ==============================================================================
# SECTION 7: CLUSTERED STANDARD ERRORS
# ==============================================================================

# Groups= takes a Series or array of cluster IDs aligned with the model data.
# Use the model's index to align properly.

ces_cc = ces.dropna(subset=["imm_restrict", "education", "age", "ideology5", "sex"])

m_base = smf.ols(
    "imm_restrict ~ education + age + ideology5 + C(sex)",
    data=ces_cc
)

m_clustered = m_base.fit(
    cov_type="cluster",
    cov_kwds={"groups": ces_cc["state_fips"]}
)

print(m_clustered.summary())


# ==============================================================================
# SECTION 8: STANDARDIZED (BETA) COEFFICIENTS
# ==============================================================================

# Standardize continuous predictors before regressing.
# Coefficients are then in SD units — comparable across predictors.

def standardize(series):
    return (series - series.mean()) / series.std()

ces_std = ces.copy()
for col in ["imm_restrict", "education", "age", "ideology5", "party_id7"]:
    ces_std[f"z_{col}"] = standardize(ces[col])

m_std = smf.ols(
    "z_imm_restrict ~ z_education + z_age + z_ideology5 + z_party_id7 + C(sex)",
    data=ces_std
).fit(cov_type="HC3")

print("\nStandardized coefficients:")
print(m_std.params.round(4))


# ==============================================================================
# SECTION 9: POST-ESTIMATION DIAGNOSTICS
# ==============================================================================

m_diag = smf.ols(
    "imm_restrict ~ education + age + ideology5 + party_id7 + C(sex) + C(census_region)",
    data=ces
).fit()

fitted   = m_diag.fittedvalues
residuals = m_diag.resid
std_resid = m_diag.get_influence().resid_studentized_internal

# --- Residuals vs. Fitted ---
fig, axes = plt.subplots(1, 2, figsize=(12, 5))

axes[0].scatter(fitted, residuals, alpha=0.1, s=2, color="navy")
axes[0].axhline(0, color="red", linestyle="--")
axes[0].set_xlabel("Fitted Values")
axes[0].set_ylabel("Residuals")
axes[0].set_title("Residuals vs. Fitted")

# --- Q-Q Plot ---
stats.probplot(std_resid, dist="norm", plot=axes[1])
axes[1].set_title("Q-Q Plot of Standardized Residuals")

plt.tight_layout()
plt.savefig("diagnostics_ols.png", dpi=300)
plt.close()

# --- Cook's Distance ---
influence = m_diag.get_influence()
cooks_d   = influence.cooks_distance[0]
n         = m_diag.nobs
k         = m_diag.df_model

hi_cook = (cooks_d > 4 / n).sum()
print(f"\nHigh Cook's D (> 4/N): {hi_cook} observations")

# --- VIF (multicollinearity) ---
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Build design matrix without intercept for VIF calculation
X = m_diag.model.exog
vif_data = pd.DataFrame({
    "feature": m_diag.model.exog_names,
    "VIF":     [variance_inflation_factor(X, i) for i in range(X.shape[1])]
}).round(2)
print("\nVIF:\n", vif_data)

# --- Breusch-Pagan test (heteroskedasticity) ---
from statsmodels.stats.diagnostic import het_breuschpagan

bp_stat, bp_p, _, _ = het_breuschpagan(m_diag.resid, m_diag.model.exog)
print(f"\nBreusch-Pagan: stat={bp_stat:.3f}, p={bp_p:.4f}")
if bp_p < 0.05:
    print("  → Heteroskedasticity detected — use robust SEs")


# ==============================================================================
# SECTION 10: FORMATTED RESULTS TABLE
# ==============================================================================

# Build a clean side-by-side coefficient table manually
def model_table(models, model_names, sig_levels=(.05, .01, .001)):
    rows = {}
    for name, m in zip(model_names, models):
        col = {}
        for param in m.params.index:
            coef = m.params[param]
            se   = m.bse[param]
            p    = m.pvalues[param]
            stars = (
                "***" if p < sig_levels[2] else
                "**"  if p < sig_levels[1] else
                "*"   if p < sig_levels[0] else ""
            )
            col[param] = f"{coef:.3f}{stars}"
            col[f"{param}_se"] = f"({se:.3f})"
        col["N"]   = int(m.nobs)
        col["R²"]  = round(m.rsquared, 3)
        col["Adj R²"] = round(m.rsquared_adj, 3)
        rows[name] = col
    return pd.DataFrame(rows)

m1 = smf.ols("imm_restrict ~ education", data=ces).fit()
m2 = smf.ols("imm_restrict ~ education + age + C(sex) + C(census_region)", data=ces).fit()
m3 = smf.ols("imm_restrict ~ education + age + C(sex) + C(census_region) + ideology5 + party_id7", data=ces).fit()

table = model_table([m1, m2, m3], ["Bivariate", "Demographics", "Full"])
print("\nRegression Table:\n", table)

# Save to CSV for import into Word
table.to_csv("table_ols_restrict.csv")

print("\nModule 3 complete.")

File: code-library/03-ols-regression/ols-regression.R
# ==============================================================================
#  MODULE 3: OLS LINEAR REGRESSION
#  Dataset: CES 2020 (cleaned in Module 1)
#
#  Purpose: Run OLS from bivariate through full models with interaction terms,
#  robust SEs, post-estimation diagnostics, and formatted output tables.
# ==============================================================================

library(tidyverse)
library(sandwich)    # robust/clustered standard errors (vcovHC, vcovCL)
library(lmtest)      # coeftest() — apply robust SEs to model output
library(modelsummary) # publication-quality regression tables
library(broom)       # tidy(), glance(), augment() — tidy model output

# install.packages(c("sandwich", "lmtest", "modelsummary", "broom"))

ces <- readRDS("CES2020_clean.rds")

# Complete-case subset spanning all variables used in the comparison table
# (Sections 1–3) so modelsummary sees the same N across nested models
ces_cc <- ces |>
  tidyr::drop_na(imm_restrict, education, age, sex, census_region,
                 ideology5, party_id7, college)


# ==============================================================================
# SECTION 1: SIMPLE (BIVARIATE) OLS
# ==============================================================================

# lm(outcome ~ predictor(s), data)
# R's formula interface: ~ separates outcome from predictors
# + adds predictors; * adds an interaction AND both main effects

m_simple <- lm(imm_restrict ~ education, data = ces_cc)

summary(m_simple)

# tidy() from broom gives a clean tibble of coefficients
tidy(m_simple, conf.int = TRUE)

# glance() gives model-level stats (R², F-stat, df, etc.)
glance(m_simple)

# Interpretation: the "estimate" for education is the slope —
# each additional unit of education is associated with that many
# fewer/more units of immigration restrictionism, on average.


# ==============================================================================
# SECTION 2: MULTIPLE REGRESSION (ADDITIVE CONTROLS)
# ==============================================================================

# factor() converts a numeric variable to a categorical dummy set in the model
# R automatically omits the lowest-level category as the reference group
# (equivalent to Stata's i. prefix)

m_demog <- lm(
  imm_restrict ~ education + age + factor(sex) + factor(census_region),
  data = ces_cc
)

summary(m_demog)
tidy(m_demog, conf.int = TRUE)


# ==============================================================================
# SECTION 3: FULL MODEL
# ==============================================================================

m_full <- lm(
  imm_restrict ~ education + age + factor(sex) + factor(census_region) +
    ideology5 + party_id7 + college,
  data = ces_cc
)

summary(m_full)
tidy(m_full, conf.int = TRUE)
glance(m_full)


# ==============================================================================
# SECTION 4: COMPARING NESTED MODELS
# ==============================================================================

# anova(restricted_model, full_model) runs an F-test comparing nested models
# (the R equivalent of Stata's "test ideology5 party_id7" after regress)

anova(m_demog, m_full)

# The F-statistic tests H0: all newly added coefficients == 0 simultaneously.
# Significant p-value → the full model fits significantly better.


# ==============================================================================
# SECTION 5: INTERACTION TERMS
# ==============================================================================

# In R formulas:
#   a:b    = interaction of a and b (no main effects)
#   a*b    = a + b + a:b (main effects + interaction)
# Always use * when you want the interaction — it includes main effects.

# Does education's effect on restrictionism vary by party?
m_interact <- lm(
  imm_restrict ~ education * factor(party_id3) + age + factor(sex),
  data = ces
)

summary(m_interact)

# Visualize the interaction with predicted values
# Create a prediction grid covering education × party combinations
pred_grid <- expand.grid(
  education  = 1:6,
  party_id3  = 1:3,
  age        = mean(ces$age, na.rm = TRUE),
  sex        = 1   # hold sex constant at Male (reference)
)

pred_grid$predicted <- predict(m_interact, newdata = pred_grid)

ggplot(pred_grid, aes(x = education, y = predicted,
                      color = factor(party_id3),
                      group = factor(party_id3))) +
  geom_line(linewidth = 1) +
  geom_point(size = 2) +
  scale_color_manual(
    values = c("blue", "red", "darkgreen"),
    labels = c("Democrat", "Republican", "Independent")
  ) +
  scale_x_continuous(
    breaks = 1:6,
    labels = c("No HS", "HS grad", "Some coll.", "Assoc.", "Bach.", "Postgrad")
  ) +
  labs(
    title   = "Predicted Immigration Restrictionism by Education & Party",
    x       = "Education Level",
    y       = "Predicted Restrictionism (0–5)",
    color   = "Party ID"
  ) +
  theme_minimal()


# ==============================================================================
# SECTION 6: ROBUST STANDARD ERRORS (HETEROSKEDASTICITY-ROBUST)
# ==============================================================================

# OLS standard errors assume homoskedasticity. When violated, use
# Huber-White robust SEs via sandwich::vcovHC() and lmtest::coeftest().
#
# type = "HC3" is the most common; "HC1" matches Stata's default robust

m_robust <- lm(
  imm_restrict ~ education + age + ideology5 + factor(sex) + factor(census_region),
  data = ces
)

# Display with robust SEs
coeftest(m_robust, vcov = vcovHC(m_robust, type = "HC3"))

# Tidy output with robust SEs — use tidy(coeftest(...))
tidy(coeftest(m_robust, vcov = vcovHC(m_robust, type = "HC3")), conf.int = TRUE)


# ==============================================================================
# SECTION 7: CLUSTERED STANDARD ERRORS
# ==============================================================================

# Cluster SEs by state when observations within states are correlated.
# vcovCL() from sandwich handles clustering.

m_cluster <- lm(
  imm_restrict ~ education + age + ideology5 + factor(sex),
  data = ces
)

coeftest(m_cluster, vcov = vcovCL(m_cluster, cluster = ~state_fips))


# ==============================================================================
# SECTION 8: STANDARDIZED (BETA) COEFFICIENTS
# ==============================================================================

# Standardize predictors to SD units before regressing.
# scale() centers and divides by SD (z-score).
# Resulting coefficients are directly comparable in magnitude.

ces_scaled <- ces |>
  mutate(across(
    c(education, age, ideology5, party_id7, imm_restrict),
    ~ as.numeric(scale(.x))
  ))

m_std <- lm(
  imm_restrict ~ education + age + ideology5 + party_id7 + factor(sex),
  data = ces_scaled
)

tidy(m_std, conf.int = TRUE)

# All coefficients are now in SD units — compare magnitude across predictors.
# factor(sex) is a dummy so its standardized coefficient is less interpretable.


# ==============================================================================
# SECTION 9: POST-ESTIMATION DIAGNOSTICS
# ==============================================================================

# augment() from broom adds fitted values, residuals, leverage, Cook's D
# to the original data as new columns

diag_df <- augment(m_full)

# --- Residual vs. Fitted ---
ggplot(diag_df, aes(x = .fitted, y = .resid)) +
  geom_point(alpha = 0.2, color = "navy", size = .5) +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  geom_smooth(method = "loess", se = FALSE, color = "orange", linewidth = .8) +
  labs(title = "Residuals vs. Fitted Values",
       x = "Fitted Values", y = "Residuals") +
  theme_minimal()

# --- Q-Q Plot (normality of residuals) ---
ggplot(diag_df, aes(sample = .std.resid)) +
  stat_qq(alpha = 0.3, color = "navy") +
  stat_qq_line(color = "red") +
  labs(title = "Normal Q-Q Plot of Standardized Residuals",
       x = "Theoretical Quantiles", y = "Standardized Residuals") +
  theme_minimal()

# --- Cook's D — influential observations ---
diag_df |>
  mutate(obs = row_number()) |>
  ggplot(aes(x = obs, y = .cooksd)) +
  geom_point(alpha = 0.3, size = .5) +
  geom_hline(yintercept = 4 / nrow(ces), color = "red", linetype = "dashed") +
  labs(title = "Cook's Distance",
       subtitle = "Red line = 4/N threshold",
       x = "Observation", y = "Cook's D") +
  theme_minimal()

# Count high-influence observations
n <- nrow(m_full$model)
sum(diag_df$.cooksd > 4 / n, na.rm = TRUE)

# --- VIF (multicollinearity) ---
# install.packages("car")
library(car)
vif(m_full)
# VIF > 10 (or > 5 conservatively) signals problematic multicollinearity

# --- Breusch-Pagan test (heteroskedasticity) ---
# lmtest::bptest()
bptest(m_full)
# Significant → heteroskedasticity present → use robust SEs


# ==============================================================================
# SECTION 10: PUBLICATION TABLE WITH MODELSUMMARY
# ==============================================================================

# modelsummary() produces Word, LaTeX, or HTML regression tables.
# Pass a named list of models for side-by-side columns.

models <- list(
  "Bivariate"    = m_simple,
  "Demographics" = m_demog,
  "Full Model"   = m_full
)

# Console display
modelsummary(
  models,
  stars   = c("*" = .05, "**" = .01, "***" = .001),
  gof_map = c("nobs", "r.squared", "adj.r.squared", "AIC", "BIC"),
  title   = "OLS Regression: Predictors of Immigration Restrictionism"
)

# Export to Word
# modelsummary(models, output = "table_ols.docx",
#   stars = c("*" = .05, "**" = .01, "***" = .001))

# Export to LaTeX
# modelsummary(models, output = "table_ols.tex",
#   stars = c("*" = .05, "**" = .01, "***" = .001))

# With robust SEs
modelsummary(
  models,
  vcov  = "HC3",   # applies robust SEs to all models automatically
  stars = c("*" = .05, "**" = .01, "***" = .001),
  title = "OLS Regression: Robust SEs"
)

message("Module 3 complete.")
